airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -173,14 +173,11 @@ def _get_concurrent_state(self, state: MutableMapping[str, Any]) -> Tuple[Cursor
         return self._connector_state_converter.convert_from_sequential_state(self._cursor_field, state, self._start)
 
     def observe(self, record: Record) -> None:
-        if self._slice_boundary_fields:
-            # Given that slicing is done using the cursor field, we don't need to observe the record as we assume slices will describe what
-            # has been emitted. Assuming there is a chance that records might not be yet populated for the most recent slice, use a lookback
-            # window
-            return
+        most_recent_cursor_value = self._most_recent_cursor_value_per_partition.get(record.partition)
```

Comment:
This code is not thread safe but `cursor.observe` is called in `StreamPartition` which is executed in a thread. Even the previous code seems at risk issue here (although less than the new one as it would only affect the cases where there are no cursor boundaries. Should we either move this to the main thread or have a lock here? I would prefer the main thread solution as it would mean stream state is decoupled from the stream but it feels like a bigger change. @brianjlai I'd love your opinion on this

Reply:
Presumably you mean because we call `observe()` from within the adapter when we're performing a read of a single partition? code here: https://github.com/airbytehq/airbyte/blob/4f6c29a9ca04b0c69a9d614fb344abf69ffc6333/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py#L259

It seems like at a glance we could move the call to `cursor.observe` out into the `ConcurrentReadProcessor.on_record`. I'd probably prefer that over the lock as well.

@maxi297 is the part that feels like a bigger change because we need to know which cursor to call observe on from within `on_record`? I played around with `ConcurrentReadProcessor`, and i think we can derive the stream and cursor from one of the fields instantiated `self._stream_name_to_instance`. Is that what you're thinking?

Reply:
> It seems like at a glance we could move the call to cursor.observe out into the ConcurrentReadProcessor.on_record. I'd probably prefer that over the lock as well.

Exactly!

> is the part that feels like a bigger change because we need to know which cursor to call observe on from within on_record?

I didn't remember that we had the cursor as part of the stream interface. I think we can just call `stream.cursor.observe` as part of `on_record` and that's it. That would be a much smaller change that I expected